timemachine/ff/handlers/utils.py

```diff
@@ -67,3 +103,219 @@ def match_smirks(mol, smirks):
+    charges from the template_res_mol.
+
+    Parameters
+    ----------
+    topology_res_mol:
+        Generated using `make_residue_mol` to generate an RDKit mol
+        from an OpenMM residue. This molecule is updated in place
+        with the bond types and formal charges from `template_res_mol`.
+    template_res_mol:
+        Template molecule generated using `make_residue_mol_from_template`.
+    """
+    match = template_res_mol.GetSubstructMatch(get_query_mol(topology_res_mol))
+
+    # Match maps the topology_res_mol to template_res_mol,
+    # which has the proper bond types and charges
+    fwd_map = {i: v for i, v in enumerate(match)}
+    template_res_atoms = {i: atom for i, atom in enumerate(template_res_mol.GetAtoms())}
+    template_res_mol.UpdatePropertyCache()
+
+    for atom in topology_res_mol.GetAtoms():
+        template_res_idx = fwd_map[atom.GetIdx()]
+        template_res_atom = template_res_atoms[template_res_idx]
+
+        if template_res_atom.GetFormalCharge() != 0:
+            atom.SetFormalCharge(template_res_atom.GetFormalCharge())
+
+    # Update the bonds/aromatic type based on the template
+    template_bonds = {}
+    for bond in template_res_mol.GetBonds():
+        src_idx = bond.GetBeginAtomIdx()
+        dst_idx = bond.GetEndAtomIdx()
+        template_bonds[(src_idx, dst_idx)] = bond
+        template_bonds[(dst_idx, src_idx)] = bond
+
+    topology_res_mol.UpdatePropertyCache()
+    for bond in topology_res_mol.GetBonds():
+        src_idx = bond.GetBeginAtomIdx()
+        dst_idx = bond.GetEndAtomIdx()
+
+        template_src_idx = fwd_map[src_idx]
+        template_dst_idx = fwd_map[dst_idx]
+        k = (template_src_idx, template_dst_idx)
+        bond.SetBondType(template_bonds[k].GetBondType())
+        bond.SetIsAromatic(template_bonds[k].GetIsAromatic())
+
+    topology_res_mol.UpdatePropertyCache()
+
+
+def get_res_name(template_name: str) -> Tuple[str, bool, bool]:
+    """
+    Parameters
+    ----------
+    template_name:
+        Three code of the residue.
+        Can have an 'N' or 'C' prefix to indicate a capped residue.
+
+    Return
+    ------
+    res_name: str
+        3 letter residue name
+    has_n_cap: bool
+        True if this residue is capped at the N terminus.
+    has_c_cap: bool
+        True if this residue is capped at the C terminus.
+    """
+    has_c_cap = False
+    has_n_cap = False
+    res_name = template_name
+    if len(res_name) == 4 and res_name[0] == "C":
+        res_name = res_name[1:]
+        has_c_cap = True
+    if len(res_name) == 4 and res_name[0] == "N":
+        has_n_cap = True
+        res_name = res_name[1:]
+
+    assert len(res_name) == 3, f"Invalid residue name: {template_name}"
+    return res_name, has_n_cap, has_c_cap
+
+
+def add_n_cap(mol: Mol) -> Mol:
+    """
+    Add an N cap to the template residue mol.
+
+    NOTE: This only works properly for residues constructed
+    from the template, see `SMILES_BY_RES_NAME`.
```

Comment:
Usage note is sufficient for now, but it may also be useful to assert that (canonicalized) MolToSmiles(mol) is in SMILES_BY_RES_NAME or similar, here and in `add_c_cap`